src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java

```diff
@@ -378,6 +401,253 @@ public QuizExerciseResource(QuizExerciseRepository quizExerciseRepository, Parti
+            updateOfResultsAndStatisticsNecessary = true;
+        }
+        return updateOfResultsAndStatisticsNecessary;
+    }
+
+    /**
+     * 1. undo all changes which are not allowed ( adding Answers)
+     * 2. check if an update of the Results and Statistics is necessary
+     *
+     * @param question the changed MultipleChoiceQuestion-object
+     * @param originalQuestion the original MultipleChoiceQuestion-object, which will be compared with the new changed question
+     *
+     * @return a boolean which is true if the answer-changes make an update necessary and false if not
+     */
+    private boolean undoUnallowedChangesAndCheckMultipleChoiceQuestionAnswers (MultipleChoiceQuestion question, MultipleChoiceQuestion originalQuestion){
+
+        boolean updateNecessary = false;
+
+        //find added Answers, which are not allowed to be added
+        Set<AnswerOption> notAllowedAddedAnswers = new HashSet<>();
+        //check every answer of the question
+        for (AnswerOption answer : question.getAnswerOptions()) {
+            //check if the answer were already in the originalQuizExercise -> if not it's an added answer
+            if (originalQuestion.getAnswerOptions().contains(answer)) {
+                //find original answer
+                AnswerOption originalAnswer = originalQuestion.findAnswerOptionById(answer.getId());
+                //reset invalid answer if it already set to true (it's not possible to set an answer valid again)
+                answer.setInvalid(answer.isInvalid() || originalAnswer.isInvalid());
+
+                // check if an answer is set invalid or if the correctness has changed
+                // if true an update of the Statistics and Results is necessary
+                if ((answer.isInvalid() && !originalAnswer.isInvalid() && !question.isInvalid()) ||
+                    (!(answer.isIsCorrect().equals(originalAnswer.isIsCorrect())))) {
+                    updateNecessary = true;
+                }
+            } else {
+                //mark the added Answers (adding questions is not allowed)
+                notAllowedAddedAnswers.add(answer);
+            }
+        }
+        //remove the added Answers
+        question.getAnswerOptions().removeAll(notAllowedAddedAnswers);
+
+        // check if an answer was deleted (not allowed added answers are not relevant)
+        // if true an update of the Statistics and Results is necessary
+        if ( question.getAnswerOptions().size() < originalQuestion.getAnswerOptions().size()) {
+            updateNecessary = true;
+        }
+        return updateNecessary;
+    }
+
+    /**
+     * 1. Go through all Results in the Participation and recalculate the score
+     * 2. recalculate the statistic the given quizExercise
+     *
+     * @param quizExercise the changed QuizExercise object which will be used to recalculate the existing Results and Statistics
+     */
```

Comment:
This method is too long. Please split it into multiple smaller methods and rather use a Service or move methods to the corresponding domain classes instead of doing all the work in the resource. Again: a method shall not have more than 15-20 lines and shall have one small purpose.

Please also avoid to cast objects multiple times such as (QuizSubmission) result.getSubmission() --> it is better to cast them safely once and then use the casted object in all subsequent occurrences